[WIP] Add (optional) extended dashboards in ScyllaDBMonitoring and drop the unmanaged setup (#1161)
scylladb/scylla-operator: docs/source/monitoring.md

# Monitoring

Scylla Operator 1.8 introduced a new API object `ScyllaDBMonitoring`, allowing users to deploy a managed monitoring 
setup for their Scylla Clusters.

```yaml
apiVersion: scylla.scylladb.com/v1alpha1
kind: ScyllaDBMonitoring
metadata:
  name: example
spec:
  type: Platform
  endpointsSelector:
    matchLabels:
      app.kubernetes.io/name: scylla
      scylla-operator.scylladb.com/scylla-service-type: identity
      scylla/cluster: replace-with-your-scyllacluster-name
  components:
    prometheus:
      storage:
        volumeClaimTemplate:
          spec:
            resources:
              requests:
                storage: 1Gi
    grafana:
      exposeOptions:
        webInterface:
          ingress:
            ingressClassName: haproxy
            dnsDomains:
            - test-grafana.test.svc.cluster.local
            annotations:
              haproxy-ingress.github.io/ssl-passthrough: "true"
```

For details, refer to the below command:
```console
$ kubectl explain scylladbmonitorings.scylla.scylladb.com/v1alpha1
```

## Deploy managed monitoring

> **Note**: as of v1.8, ScyllaDBMonitoring is in alpha state and is susceptible to backwards incompatible changes.

Reviewer: ```suggestion
**Note**: as of v1.8, ScyllaDBMonitoring is experimental. The API is currently at v1alpha1 version and may change in the future versions.
```

### Requirements

> Before you can set up your ScyllaDB Monitoring, you must install the Scylla Operator in your Kubernetes cluster 

Reviewer: ```suggestion
Before you can set up your ScyllaDB Monitoring, you have to have Scylla Operator already installed in your Kubernetes cluster
```

> and have an instance of ScyllaCluster running. For details see:

Reviewer: i don't think that's technically true

Reviewer: You can find more information on how to deploy Scylla Operator here:

* [Deploying Scylla on a Kubernetes Cluster](generic.md)
* [Deploying Scylla stack using Helm Charts](helm.md)

> The managed monitoring setup also requires HAProxy Ingress and Prometheus Operator deployed in your Kubernetes cluster.

Reviewer: it's not required by the monitoring itself, it's just that the example uses it

You can deploy them using the provided third party examples.

#### Deploy Prometheus Operator

> Deploy Prometheus Operator using kubectl:

Reviewer: if your cluster already has a prometheus operator deployed you should skip this step

Author: I've put a message above saying you can skip these steps if you already have them deployed in your cluster.

```console
$ kubectl -n prometheus-operator apply --server-side -f examples/third-party/prometheus-operator
```

##### Verify Prometheus Operator status
```console

> $ kubectl -n prometheus-operator get deployments.apps/prometheus-operator

Reviewer: kubectl rollout status

NAME                  READY   UP-TO-DATE   AVAILABLE   AGE
prometheus-operator   1/1     1            1           5s
```

#### Deploy HAProxy Ingress

> Deploy HAProxy Ingress using kubectl:

Reviewer: same skip here

```console
$ kubectl -n haproxy-ingress apply --server-side -f examples/third-party/haproxy-ingress
```

##### Verify HAProxy Ingress status
```console

> $ kubectl -n haproxy-ingress get deployments.apps/haproxy-ingress

Reviewer: rollout status

NAME              READY   UP-TO-DATE   AVAILABLE   AGE
haproxy-ingress   1/1     1            1           5s
```

Before you begin, update the `endpointSelector` in `examples/monitoring/v1alpha1/scylladbmonitoring.yaml` with a label 

> matching your ScyllaCluster instance.

Reviewer: instance name.

### Deploy ScyllaDBMonitoring

Deploy the monitoring setup using kubectl:
```console

> $ kubectl apply --server-side -f examples/monitoring/v1alpha1/scylladbmonitoring.yaml

Reviewer: ```suggestion
$ kubectl apply --server-side -f ./examples/monitoring/v1alpha1/scylladbmonitoring.yaml
```

```

> The Operator should pick up the object's creation and create all the necessary resources for you.

Reviewer: ```suggestion
Scylla Operator will notice the new ScyllaDBMonitoring object and create and manage all the necessary resources for you.
```

#### Verify ScyllaDBMonitoring status
```console

> $ kubectl get ScyllaDBMonitoring example

Reviewer: wait for conditions

Author: There are multiple so I had to use a workaround

Author: Also asked on https://github.com/kubernetes/kubernetes/issues/95759 if I can give fixing this a try

NAME      AGE
example   60s
```

#### Verify Prometheus status
```console
$ kubectl get statefulset.apps/prometheus-example
NAME                 READY   AGE
prometheus-example   1/1     68s
```

#### Verify Grafana status
```console
$ kubectl get deployments.apps/example-grafana
NAME              READY   UP-TO-DATE   AVAILABLE   AGE
example-grafana   1/1     1            1           68s
```

#### Verify Grafana admin credentials secret
Among other resources, the Operator creates a secret containing the admin credentials for you Grafana deployment.
```console
$ kubectl get secrets/example-grafana-admin-credentials
NAME                                TYPE     DATA   AGE
example-grafana-admin-credentials   Opaque   2      68s
```

> ### Access Grafana locally

Reviewer: I think this should be a variant when you don't want to use/install haproxy, or expose it to the world

Author: Do you suggest removing it or adding it as an alternative?

Reviewer: I think we'd primarily describe a way to use the ingress, this would be an alternative, if needed

* [Deploying Scylla on a Kubernetes Cluster](generic.md)
* [Deploying Scylla stack using Helm Charts](helm.md)

The above example of the monitoring setup also makes use of HAProxy Ingress and Prometheus Operator.
You can deploy them in your Kubernetes cluster using the provided third party examples. If you already have them deployed
in your cluster, you can skip the below steps.

#### Deploy Prometheus Operator
Deploy Prometheus Operator using kubectl:
```console
$ kubectl -n prometheus-operator apply --server-side -f ./examples/third-party/prometheus-operator
```

##### Verify Prometheus Operator status
```console
$ kubectl -n prometheus-operator rollout status --timeout=5m deployments.apps/prometheus-operator
deployment "prometheus-operator" successfully rolled out
```

#### Deploy HAProxy Ingress
Deploy HAProxy Ingress using kubectl:
```console
$ kubectl -n haproxy-ingress apply --server-side -f ./examples/third-party/haproxy-ingress
```

##### Verify HAProxy Ingress status
```console
$ kubectl -n haproxy-ingress rollout status --timeout=5m deployments.apps/haproxy-ingress
deployment "haproxy-ingress" successfully rolled out
```

> Before you begin, update the `endpointSelector` in `examples/monitoring/v1alpha1/scylladbmonitoring.yaml` with a label 

Reviewer 2: The code uses plural form of the first word -> `endpointsSelector`.

matching your ScyllaCluster instance name.

### Deploy ScyllaDBMonitoring

Deploy the monitoring setup using kubectl:
```console
$ kubectl apply --server-side -f ./examples/monitoring/v1alpha1/scylladbmonitoring.yaml
```

Scylla Operator will notice the new ScyllaDBMonitoring object, and it will reconcile all necessary resources.

#### Verify ScyllaDBMonitoring status
```console
$ kubectl wait --for='condition=Available=True' scylladbmonitorings.scylla.scylladb.com/example
scylladbmonitoring.scylla.scylladb.com/example condition met

> $ kubectl wait --for='condition=Progressing=False' scylladbmonitorings.scylla.scylladb.com/example

Reviewer: with this raciness, you should check progressing first and available last. it can be still available for a while when it starts progressing and when it rolls out you won't check it again

Author: good point, updated

$ kubectl wait --for='condition=Available=True' scylladbmonitorings.scylla.scylladb.com/example && \
kubectl wait --for='condition=Progressing=False' scylladbmonitorings.scylla.scylladb.com/example && \
kubectl wait --for='condition=Degraded=False' scylladbmonitorings.scylla.scylladb.com/example
scylladbmonitoring.scylla.scylladb.com/example condition met
scylladbmonitoring.scylla.scylladb.com/example condition met
scylladbmonitoring.scylla.scylladb.com/example condition met
```

#### Verify Prometheus status
```console
$ kubectl rollout status --timeout=5m statefulset.apps/prometheus-example
statefulset rolling update complete 1 pods at revision prometheus-example-65b89d55bb...
```

#### Verify Grafana status
```console
$ kubectl rollout status --timeout=5m deployments.apps/example-grafana
deployment "example-grafana" successfully rolled out
```

#### Verify Grafana admin credentials secret
Among other resources, the Operator creates a secret containing the admin credentials for you Grafana deployment.
```console
$ kubectl get secrets/example-grafana-admin-credentials
NAME                                TYPE     DATA   AGE
example-grafana-admin-credentials   Opaque   2      68s

> ```

Author: Yeah it doesn't really make sense, I left it there to demonstrate how to access the admin credentials. Changed.

> ### Access Grafana

Reviewer: I feel that this article is too focused on how to hack it with minikube and someone trying to set this up on a real cluster / domain would be confused.

I think the generic steps are:
- create ScyllaDBMonitoring with a domain (`test-grafana.test.svc.cluster.local` is fine)
- [real domain / cname to ingress controller] curl
- [fake domain / port-forward / not a cname to ingress controller] curl + `--resolve`

Author: > - [real domain / cname to ingress controller] curl
> - [fake domain / port-forward / not a cname to ingress controller] curl + --resolve

I don't understand these lines. Could you elaborate on what these options in the brackets mean? Do you mean that there should be two sections demonstrating all of these options, verifying them with either curl or curl --resolve? Are you also suggesting not using minikube to demonstrate it?

Reviewer: I think that could be a joint section saying if your scylldbmonitoring domain isn't resolvable (to the ingress controller IP accessible from outside of the cluster), then you have to add the --resolve flag with ingress controller IP or use `/etc/hosts` alias. The ingress controller IP (accessible from outside of the cluster) can be constructed in a few ways - as service externalIP, nodeIP +  service nodePort, or 127.0.0.1+kubectl port-forward to ingress controller svc, ...

#### Wait for ScyllaDBMonitoring to roll out
```console
$ kubectl wait --for='condition=Progressing=False' scylladbmonitorings.scylla.scylladb.com/example
scylladbmonitoring.scylla.scylladb.com/example condition met

$ kubectl wait --for='condition=Degraded=False' scylladbmonitorings.scylla.scylladb.com/example
scylladbmonitoring.scylla.scylladb.com/example condition met

$ kubectl wait --for='condition=Available=True' scylladbmonitorings.scylla.scylladb.com/example
scylladbmonitoring.scylla.scylladb.com/example condition met
```

#### Wait for Prometheus to roll out
```console
$ kubectl rollout status --timeout=5m statefulset.apps/prometheus-example
statefulset rolling update complete 1 pods at revision prometheus-example-65b89d55bb...
```

#### Wait for Grafana to roll out
```console
$ kubectl rollout status --timeout=5m deployments.apps/example-grafana
deployment "example-grafana" successfully rolled out
```

### Accessing Grafana

For accessing Grafana service from outside the Kubernetes cluster we recommend using an Ingress, although there are many other ways to do so.

> When using Ingress, what matters is to direct your packets to the ingress controller Service/Pods and have the correct TLS SNI field set by the caller when reaching out to the service, so it is routed properly, and it can successfully validate the grafana serving certificate. 

Reviewer: ```suggestion
When using Ingress, what matters is to direct your packets to the ingress controller Service/Pods and have the correct TLS SNI field set by the caller when reaching out to the service, so it is routed properly, and your client can successfully validate the grafana serving certificate.
```

Note: To test a resolvable domain from your machine without creating DNS records, you can adjust `/etc/hosts` or similar.

```console
$ curl --fail -s -o /dev/null -w '%{http_code}' -L --cacert <( echo "${GRAFANA_SERVING_CERT}" ) "https://example-grafana.apps.mydomain" --user "${GRAFANA_USER}:${GRAFANA_PASSWORD}"
200
```

### Connecting through Ingress using an unresolvable domain

To connect to an Ingress without a resolvable domain you first need to find out your Ingress controller's IP that can be resolved externally. Again, there are many ways to do so beyond the below examples.

Unless stated otherwise, we assume your Ingress is running on port 443.

```console
$ INGRESS_PORT=443
```

#### Variants

##### Ingress ExternalIP

When you are running in a real cluster there is usually a cloud LoadBalancer or a bare metal alternative providing you with an externally reachable IP address.

```console
$ INGRESS_IP="$( kubectl -n=haproxy-ingress get service/haproxy-ingress --template='{{ ( index .loadBalancer.ingress 0 ).ip }}' )"
```

##### Ingress NodePort

NodePort is slightly less convenient, but it's available in development clusters as well.

```console

> $ INGRESS_IP="$( kubectl get nodes --template='{{ $internal_ip := "" }}{{ $external_ip := "" }}{{ range ( index .items 0 ).status.addresses }}{{ if eq .type "InternalIP" }}{{ $internal_ip = .address }}{{ else if eq .type "ExternalIP" }}{{ $external_ip = .address }}{{ end }}{{ end }}{{ if $external_ip }}{{ $external_ip }}{{ else }}{{ $internal_ip }}{{ end }}' )"

Reviewer: Shouldn't we prefer the `ExternalIP` first and `InternalIP` as a fallback as it work only if you are in the same network like with your VM but not for say a cloud? The other way around both should work with a fallback.

Reviewer: you are right, I've misread while scrolling

HTTP/1.1 200 OK
Cache-Control: no-cache
Content-Type: text/html; charset=UTF-8
Expires: -1
Pragma: no-cache
X-Content-Type-Options: nosniff
X-Frame-Options: deny
X-Xss-Protection: 1; mode=block
Date: Fri, 07 Apr 2023 14:17:37 GMT
```

###### Port forward

Alternatively, you can use `kubcetl port-forward` to forward the traffic between the host and the Grafana service directly:

```console
$ kubectl -n scylla port-forward services/example-grafana 3000:3000
```

And then use a similar command to access Grafana again:
```console
$ curl --fail --retry-all-errors 10 --max-time 3 --retry-max-time 5 --cacert <( echo "${GRAFANA_SERVING_CERT}" ) --resolve "test-grafana.test.svc.cluster.local:3000:127.0.0.1" -IL "https://test-grafana.test.svc.cluster.local:3000" --user "${GRAFANA_USER}:${GRAFANA_PASSWORD}"

HTTP/1.1 200 OK
Cache-Control: no-cache
Content-Type: text/html; charset=UTF-8
Expires: -1
Pragma: no-cache
X-Content-Type-Options: nosniff
X-Frame-Options: deny
X-Xss-Protection: 1; mode=block
Date: Fri, 07 Apr 2023 14:32:09 GMT

> ```

Reviewer: I've tried to simplify this and also some of the templates. Please verify the instructions.

### Accessing Grafana

For accessing Grafana service from outside the Kubernetes cluster we'd recommend using an Ingress but there are many other ways. What matters when using the Ingress is to direct your packets to the ingress controller service/pods and have the correct TLS SNI field set by the caller when reaching out the service, so it is routed properly, and it can successfully validate the grafana serving certificate. Obviously this is easier when you are using a real DNS domain that resolves to your Ingress controller IP address but most of the client/tools allow to set the SNI field manually.

#### Prerequisites

For accessing Grafana, you first need to collect the serving CA and the credentials.

```console
$ GRAFANA_SERVING_CERT="$( kubectl -n scylla get secret/example-grafana-serving-ca --template '{{ index .data "tls.crt" }}' | base64 -d )"
$ GRAFANA_USER="$( kubectl -n scylla get secret/example-grafana-admin-credentials --template '{{ index .data "username" }}' | base64 -d )"
$ GRAFANA_PASSWORD="$( kubectl -n scylla get secret/example-grafana-admin-credentials --template '{{ index .data "password" }}' | base64 -d )"
```

#### Connecting through Ingress using a resolvable domain

In case of production cluster someone has already set up the Ingress controller and appropriate DNS records. Often there is a generic wildcard record like `*.app.mydomain` already pointing to the ingress controller external IP. For custom service domains, these are usually a CNAME pointing the Ingress controller A record.

Note: The ScyllaDBMonitoring example creates an Ingress object with `test-grafana.test.svc.cluster.local` DNS domain - you need to adjust this to your domain. Here it is represented as `example-grafana.apps.mydomain`.

Note: To test a resolvable domain from your machine without creating DNS records, you can adjust `/etc/hosts` or similar.

```console
$ curl --fail -s -o /dev/null -w '%{http_code}' -L --cacert <( echo "${GRAFANA_SERVING_CERT}" ) "https://example-grafana.apps.mydomain" --user "${GRAFANA_USER}:${GRAFANA_PASSWORD}"
200
```

#### Connecting through Ingress using an unresolvable domain

To connect to an ingress without a resolvable domain you first need to find out your Ingress controller IP that can be resolved externally. Again there are more ways to do it, not limited to the few examples shown here.

Unless stated otherwise, we will assume your ingress is running on port 443.
```console
$ INGRESS_PORT=443
```

##### Variants
###### Ingress ExternalIP

When you are running in a real cluster there is usually a cloud LB or a bare metal alternative providing you with an externally reachable IP.

```console
$ INGRESS_IP="$( kubectl -n=haproxy-ingress get service/haproxy-ingress --template='{{ ( index .loadBalancer.ingress 0 ).ip }}' )"
```

###### Ingress NodePort

NodePort is a bit of a less convenient but it's available in dev clusters as well.

```console
# FIXME: the template needs to prefer ExternalIP, if set. 
$ INGRESS_IP="$( kubectl get nodes --template='{{ range ( index .items 0 ).status.addresses }}{{ if eq .type "InternalIP" }}{{ .address }}{{ end }}{{ end }}' )"
$ INGRESS_PORT="$( kubectl -n=haproxy-ingress get services/haproxy-ingress --template='{{ range .spec.ports }}{{ if eq .port 443 }}{{ .nodePort }}{{ end }}{{ end }}' )"
```

###### Ingress Port Forwarding

Port forwarding may be useful, if your nodes are not reachable through a public IP, nor your Ingress, or in general for development purposes. 

```console
$ kubectl -n=haproxy-ingress port-forward services/haproxy-ingress 5000:443 >/dev/null &
$ INGRESS_IP="localhost"
$ INGRESS_PORT="5000"
```

##### Connection

```console
$ curl --fail -s -o /dev/null -w '%{http_code}' -L --cacert <( echo "${GRAFANA_SERVING_CERT}" ) "https://test-grafana.test.svc.cluster.local"  --resolve "test-grafana.test.svc.cluster.local:${INGRESS_PORT}:${INGRESS_IP}" --user "${GRAFANA_USER}:${GRAFANA_PASSWORD}"
200
```

Author: I guess we had a different perspective on how this should look like but this looks good to me - I'll vet the commands and fix the typos and other errors.
